fix default behavior of overlap objective

rajgane07 opened the pull request:
Initial Commit

rajgane07 commented:
After making changes, I get the below numbers
Energy: 7.98, Overlap: 0.96
Energy: 8.74, Overlap: 0.91
Energy: 9.24, Overlap: 0.87
Energy: 9.55, Overlap: 0.81
Energy: 9.76, Overlap: 0.78
Energy: 9.85, Overlap: 0.77
Energy: 9.87, Overlap: 0.77
Energy: 9.87, Overlap: 0.77
Energy: 9.85, Overlap: 0.77
Energy: 9.86, Overlap: 0.77
Energy: 9.90, Overlap: 0.77

@rsln-s @alex124585 - I get the above results. Is there a way to verify these numbers?

alex124585 commented:
> After making changes, I get the below numbers Energy: 7.98, Overlap: 0.96 Energy: 8.74, Overlap: 0.91 Energy: 9.24, Overlap: 0.87 Energy: 9.55, Overlap: 0.81 Energy: 9.76, Overlap: 0.78 Energy: 9.85, Overlap: 0.77 Energy: 9.87, Overlap: 0.77 Energy: 9.87, Overlap: 0.77 Energy: 9.85, Overlap: 0.77 Energy: 9.86, Overlap: 0.77 Energy: 9.90, Overlap: 0.77
> 
> @rsln-s @alex124585 - I get the above results. Is there a way to verify these numbers?

So basically you are loading max value - np.argmax(precomputed_objectives) from precomputed objectives rather then loading all the values and looping through them

alex124585 commented:
> Energy: 9.85, Overlap: 0.77
> Energy: 9.87, Overlap: 0.77
> Energy: 9.87, Overlap: 0.77
> Energy: 9.85, Overlap: 0.77
> Energy: 9.86, Overlap: 0.77
> Energy: 9.90, Overlap: 0.77

Overlap 
 

> > After making changes, I get the below numbers Energy: 7.98, Overlap: 0.96 Energy: 8.74, Overlap: 0.91 Energy: 9.24, Overlap: 0.87 Energy: 9.55, Overlap: 0.81 Energy: 9.76, Overlap: 0.78 Energy: 9.85, Overlap: 0.77 Energy: 9.87, Overlap: 0.77 Energy: 9.87, Overlap: 0.77 Energy: 9.85, Overlap: 0.77 Energy: 9.86, Overlap: 0.77 Energy: 9.90, Overlap: 0.77
> > @rsln-s @alex124585 - I get the above results. Is there a way to verify these numbers?
> 
> So basically you are loading max value - np.argmax(precomputed_objectives) from precomputed objectives rather then loading all the values and looping through them

Overlap of final QAOA state with the ground state should it be different values for different emerges? Looks like we have same 0.77 value  for higher emerges.

rsln-s commented:
Thank you, Raj! Unfortunately, I believe these numbers are incorrect. Since you're changing the downstream `get_qaoa_objective`, there are two ways test them. I recommend adding both tests.
1. Compare the output of `get_qaoa_objective` with `precomputed_merit_factors` passed in directly (see [here](https://github.com/jpmorganchase/QOKit/blob/fa7ecb1a69e180379ab7e92550bf04b66bb09799/qokit/qaoa_objective_labs.py#L190)) with the output of `get_qaoa_labs_objective` ([here](https://github.com/jpmorganchase/QOKit/blob/fa7ecb1a69e180379ab7e92550bf04b66bb09799/tests/test_objectives.py#L66)). The two should give the same result.
2. Compare the output `get_qaoa_maxcut_objective` with only `precomputed_energies` passed in, and with both `precomputed_energies` and `precomputed_optimal_bitstrings`.

rajgane07 commented:
@rsln-s @gromitsun @alex124585 
Following are the changes in this PR.
- Modified Max Cut Objective to take negative precomputed costs
- renamed precomputed-objective to precomputed costs across the project
- fixed test cases that did not take negative precomputed costs
- Started fixing relative imports to absolute imports. It is good practices to keep all imports absolute
https://peps.python.org/pep-0008/#imports

rsln-s reviewed (changes requested):
@rajgane07, please note that `from labs import get_energy_term_indices` is a _relative_ import. Example of an absolute import would be `from qokit.labs import get_energy_term_indices`. See discussion here: https://peps.python.org/pep-0328/#guido-s-decision

Moreover, the current implementation does not work properly. A simple check:
```
(py39) (base) [omniai-jupyter@ip-100-71-16-247 QOKit]$ cd ~
(py39) (base) [omniai-jupyter@ip-100-71-16-247 ~]$ ipython
Python 3.9.4 (default, Nov 13 2023, 07:56:39) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from qokit import get_qaoa_labs_objective
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[1], line 1
----> 1 from qokit import get_qaoa_labs_objective

File /opt/omniai/work/instance1/jupyter/QOKit/qokit/__init__.py:6
      1 ###############################################################################
      2 # // SPDX-License-Identifier: Apache-2.0
      3 # // Copyright : JP Morgan Chase & Co
      4 ###############################################################################
----> 6 from qaoa_objective import get_qaoa_objective
      7 from qaoa_objective_labs import get_qaoa_labs_objective

ModuleNotFoundError: No module named 'qaoa_objective'
```

I am OK with either of the following two solutions: 

1. reverting to relative imports as they were earlier (e.g.  `from .labs import get_energy_term_indices`) 
2. changing everywhere to absolute imports (i.e. calling out the top-level package and the entire path explicitly, e.g. `from qokit.labs import get_energy_term_indices`). I have a soft preference for the former, but I'm open to discussion on this. 

This will also remove the need explicit path changes in conftest. We **must** be able to get pytest to work without such shenanigans.

rajgane07 commented:
@rsln-s I have reverted old imports on the files I touched. Can you please review and approve?  I will create separate branch to test absolute imports throughly

rsln-s commented on qokit/qaoa_objective_maxcut.py:
```diff
@@ -35,7 +35,8 @@ def get_qaoa_maxcut_objective(
     precomputed_cuts : np.array
-        precomputed cuts to compute the QAOA expectation
+        precomputed cuts to compute the QAOA expectation, for maximization problem
+        send the precomputed cuts/energies as negative
```
I think this may be wrong, no? Since we flip the sign below and since `get_qaoa_objective` assumes that the optimization problem is minimization (hence the change to `precomputed_costs`), we **do not** have to pass cuts as negative. Instead, cuts are passed directly as was the case originally?

Let me know if I'm being confused.

rsln-s commented on qokit/qaoa_objective_maxcut.py:
```diff
@@ -68,12 +69,13 @@ def get_qaoa_maxcut_objective(
     else:
         parameterized_circuit = None
-
+    # Reverse the sign for computing maximizing problem
```
Language a bit ambiguous. Let me suggest the following instead: "Reverse the sign as `get_qaoa_objective` assumes that the problem is minimization"

rajgane07 commented:
@rsln-s @gromitsun - I have made changes to validates energy result comes consistent. Let me know if we need to remove qiskit simulator

rajgane07 commented:
```
import networkx as nx
import numpy as np

from qokit.maxcut import get_adjacency_matrix, maxcut_obj
from qokit.parameter_utils import get_fixed_gamma_beta
from qokit.qaoa_objective_maxcut import get_qaoa_maxcut_objective
from qokit.utils import precompute_energies

G = nx.random_regular_graph(3, 16)
f = get_qaoa_maxcut_objective(G.number_of_nodes(), 1, G=G, parameterization="gamma beta", simulator="auto")
for p in range(1, 11):
    gamma, beta = get_fixed_gamma_beta(3, p)
    print("auto", f(-1 * np.asarray(gamma), beta))

precomputed_cuts = precompute_energies(maxcut_obj, G.number_of_nodes(), w=get_adjacency_matrix(G))

f = get_qaoa_maxcut_objective(G.number_of_nodes(), 1, precomputed_cuts=precomputed_cuts, parameterization="gamma beta", simulator="auto")
for p in range(1, 11):
    print(f(*get_fixed_gamma_beta(3, p)))
```
auto -16.285479410662475
auto -17.765922558386226
auto -18.618535680427726
auto -19.248945892387393
auto -19.68386473744343
auto -19.92041677573139
auto -20.096301890336708
auto -20.249692724146197
auto -20.35306644487773
auto -20.41877407919395

-16.285479410662475
-17.765922558386226
-18.618535680427726
-19.248945892387393
-19.68386473744343
-19.92041677573139
-20.096301890336708
-20.249692724146197
-20.35306644487773
-20.41877407919395

rsln-s commented:
@alex124585 to fix or remove documentation flow, @gromitsun and @rsln-s to review from technical PoV.

rsln-s reviewed (changes requested):
Thank you for the PR!

See comments for requested change:
1. Move `gamma` sing flip into `get_fixed_gamma_beta` and `get_sk_gamma_beta`

rajgane07 commented on tests/test_qaoa_objective_maxcut.py:
```diff
@@ -0,0 +1,35 @@
+    f_precomputedcuts = get_qaoa_maxcut_objective(G.number_of_nodes(), 1, precomputed_cuts=precomputed_cuts, parameterization="gamma beta")
+    p = 1
+    gamma, beta = get_fixed_gamma_beta(3, p)
+    energy_terms = f_terms(-1 * np.asarray(gamma), beta)
```
@rsln-s  I would like to confirm if flipping the sign will have any side-effects for other use cases? for e.g. for max cut we are reversing the sign for maximization problem , if any other use case is minimization problem. Would it not affect the results?

rsln-s replied:
`get_fixed_gamma_beta` is only used for MaxCut; I don't believe there would be any side-effects.

rajgane07 replied:
@rsln-s Some of the test cases use sample_from_weighted_Shaydulin_Lotshaw_2022.json to get beta and gamma values. Shall I go ahead to change the sign in this file?
get_sk_gamma_beta is also used in Portfolio Optimization code. Changing the sign will force sign change in those code. Either we keep same and change sign in the test case or we can create 2 methods seperately to cater for minimization and maximization respectively

rsln-s replied:
@ZichangHe, do we assume minimization or maximization for QAOA objective for portfolio optimization? We should make sure it's minimization to ensure consistency

ZichangHe replied:
We assumed minimization in portfolio optimization as well. BTW, I donot think flip the sign of gamma would translate a minimal to maximal? They are not guaranteed to be symmetric?

rsln-s replied:
Would flipping the sign break portfolio opt? Do we even test portfolio opt currently?..

ZichangHe replied:
There is a test making use of `get_sk_ini ` https://github.com/jpmorganchase/QOKit/blob/4850f358bd594da8db4697725d9386f99b093dd4/tests/test_portfolio_optimization.py#L133C13-L133C13

rajgane07 replied:
@rsln-s @ZichangHe  There are test cases for Portfolio Optimization -->tests/test_portfolio_optimization.py::test_portfolio_AR this test case fails because of get_sk_gamma_beta sign switch

rajgane07 replied:
@rsln-s My suggestion for sample_from_weighted_Shaydulin_Lotshaw_2022.json this is read the file and then change the sign within test case before testing. This is will be easier.
For get_sk_gamma_beta we leave the sign as is and make the sign switch in max_cut.

rsln-s replied:
@rajgane07 I'm OK with the proposed solution.
@ZichangHe, would you mind checking what's up with the sign in portfolio opt? It's a bit surprising that even though we're minimizing, gamma is used with the wrong sign

ZichangHe replied:
If the sign in `get_sk_gamma_beta` is flipped, we can make the change by flip the `gamma_scale` here: https://github.com/jpmorganchase/QOKit/blob/4850f358bd594da8db4697725d9386f99b093dd4/qokit/portfolio_optimization.py#L274

The portfolio setup should be fine.

ZichangHe commented:
> @rsln-s @ZichangHe @gromitsun - I have made changes. Can you please review? @alex124585 - whenever you chance, can you please fix documentation workflow? it keeps failing

It looks good to me!

rsln-s reviewed (approved):
LGTM! Thanks, @rajgane07 and thanks all!
